config/runtime/base/python.jinja2

```diff
@@ -27,52 +28,79 @@ TARBALL_URL = '{{ tarball_url }}'
 WORKSPACE = '/tmp/kci'
 {%- endblock %}
 
-{% block python_body %}{% endblock %}
+{% block python_base_job -%}
+class BaseJob:
 
-{% block python_main -%}
-def _get_db():
-    if not DB_CONFIG_YAML:
-        return None
+    def __init__(self, workspace):
+        self._workspace = workspace
 
-    db_config = kernelci.config.db.DatabaseFactory.from_yaml(
-        'db', yaml.load(DB_CONFIG_YAML, Loader=yaml.CLoader)
-    )
-    if not db_config:
-        return None
+    def _get_db(self, db_config_yaml):
+        db_config = kernelci.config.db.DatabaseFactory.from_yaml(
+            'db', yaml.load(db_config_yaml, Loader=yaml.CLoader)
+        )
+        if not db_config:
+            return None
 
-    api_token = os.getenv('API_TOKEN')
-    if not api_token:
-        return None
+        api_token = os.getenv('API_TOKEN')
+        if not api_token:
+            return None
 
-    return kernelci.db.get_db(db_config, api_token)
+        return kernelci.db.get_db(db_config, api_token)
 
+    def _get_source(self, url):
+        resp = requests.get(url, stream=True)
+        resp.raise_for_status()
+        tarball_name = os.path.basename(urllib.parse.urlparse(url).path)
+        base, ext = tarball_name.split('.tar.')
+        with tarfile.open(fileobj=resp.raw, mode=f'r|{ext}') as tarball:
+            tarball.extractall(path=self._workspace)
+        return os.path.join(self._workspace, base)
 
-def _get_source(url):
-    resp = requests.get(url, stream=True)
-    resp.raise_for_status()
-    tarball_name = os.path.basename(urllib.parse.urlparse(url).path)
-    base, ext = tarball_name.split('.tar.')
```

Comment:
I'd use `os.splitext()` for getting file extenstion. It makes sure you'll get the part after the last dot. It handles the corner cases for you.